### PR TITLE
feat!: support ibis 2.x, drop ibis 1.x

### PR DESCRIPTION
**Pull request by datapythonista**
I didn't run the tests, and this won't work as is, but I implement here the main changes in the backend API for ibis-bigquery to work with ibis 2.0.

The main changes are:
- Moving all methods from `Client` to `Backend` and removing client.
- Implementing a new `Compiler` class, which is the only attribute `Backend` know about the compiler
- Remove a lot of repetitive classes in the compiler `Select`, `TableFormatter`...
- For classes that are still needed, instead of linking one another, `Compiler` has attributes to all them
- Removing the weird loop `to_sql`, `build_ast`, `Context`, `Translator`. `Compiler` contains the functionality
- Removing some repetitive methods in `Backend`, like `exists_table`, which is implemented in the base backend for all backends
- In `list_tables` and `list_databases`, the parameter `like` is now handled by a common function that implements the filtering
- Base backends have been consolidated and standardized under `ibis.backends.base`

I don't have the time, the access to bigquery or the knowledge about bigquery to leave this working, but with these main changes, and all other backends as reference it shouldn't be difficult to get this working.

For the OmniSciDB backend there is a single PR that updates the backend from Ibis 1 to Ibis 2: https://github.com/omnisci/ibis-omniscidb/pull/39

@tswast let me know if you've got any specific question or problem.

Fixes https://github.com/ibis-project/ibis-bigquery/issues/90

### CODE REVIEW

**Comment by datapythonista** [comment]
To test that this is working properly, I'd add builds to the CI that install the master Ibis, and I'd run the backend tests with `PYTEST_BACKENDS="bigquery" python -m pytest --pyargs ibis.backends.tests`. This way, besides the tests in this repo, the tests in the Ibis repo that we use for all backends will also be executed.

File: ibis_bigquery/client.py [hunk 1]
```diff
@@ -18,14 +18,10 @@
 import ibis.expr.schema as sch
 import ibis.expr.types as ir
 import pandas as pd
-import regex as re
 from google.api_core.client_info import ClientInfo
-from google.api_core.exceptions import NotFound
-from ibis.client import Database, Query, SQLClient
+from ibis.client import Database, Query
```
Comment (tswast) [inline ibis_bigquery/client.py:22]: I get

```
ibis_bigquery/client.py:22: in <module>
    from ibis.client import Database, Query
E   ModuleNotFoundError: No module named 'ibis.client'
```

when trying to run the tests locally against master. I don't see a `client` module here: https://github.com/ibis-project/ibis/tree/master/ibis Did this move?

Comment (tswast) [reply ibis_bigquery/client.py:22]: Looks like yes in https://github.com/ibis-project/ibis/pull/3035

Comment (tswast) [reply ibis_bigquery/client.py:22]: But the Query class was removed? The only references I find to it now are here: https://github.com/ibis-project/ibis/blob/6f6f0ff77899afbce694a0d25cf3edd0d297f33c/docs/source/user_guide/design.rst

Comment (datapythonista) [reply ibis_bigquery/client.py]: `ibis.client` doesn't exist, and it was moved to `ibis.backends.base`. `Query` was indeed removed, and whatever was in it was moved to `Client`, and later to `Backend`: #2789

For some backends, there is still a `client.py`, that contains `Database` and other things, but the `Client` classes don't exist and everything there have been moved to `Backend`.

**Comment by tswast** [comment]
@datapythonista I'm getting an error with the compile method from the base backend. Maybe something is missing here? https://github.com/ibis-project/ibis/blob/887efbd4c9d0657f5639638eebea53906044b78f/ibis/backends/base/__init__.py#L316

```
_______________________________________________________________________ test_large_compile _______________________________________________________________________

    def test_large_compile():
        """
        Tests that compiling a large expression tree finishes
        within a reasonable amount of time
        """
        num_columns = 20
        num_joins = 7
    
        class MockBackend(ibis_bigquery.Backend):
            def __init__(self):
                pass
    
        names = [f"col_{i}" for i in range(num_columns)]
        schema = ibis.Schema(names, ["string"] * num_columns)
        ibis_client = MockBackend()
        table = TableExpr(ops.SQLQueryResult("select * from t", schema, ibis_client))
        for _ in range(num_joins):
            table = table.mutate(dummy=ibis.literal(""))
            table = table.left_join(table, ["dummy"])[[table]]
    
        start = datetime.datetime.now()
>       table.compile()

tests/unit/test_compiler.py:314: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = ref_0
Selection[table]
  table:
    SQLQueryResult[table]
      query:
        select * from t
      schema:
        c...     right:
            dummy = Column[string*] 'dummy' from table
              ref_19
  selections:
    Table: ref_18
limit = None, timecontext = None, params = None

    def compile(
        self,
        limit=None,
        timecontext: Optional[TimeContext] = None,
        params=None,
    ):
        """
        Compile expression to whatever execution target, to verify
    
        Returns
        -------
        compiled : value or list
           query representation or list thereof
        """
>       return self._find_backend().compile(
            self, limit=limit, timecontext=timecontext, params=params
        )
E       TypeError: compile() got an unexpected keyword argument 'limit'

../ibis/ibis/expr/types.py:291: TypeError
```

**Comment by datapythonista** [comment]
I don't remember exactly, that was changed long time ago. But I think you need to use `to_ast_ensure_limit` from the compiler to set the limit, and then call the `compile` method in the ast, instead of calling the compile in the backend directly.

**Comment by jreback** [comment]
@tswast @datapythonista 

would be great to finish this up and release a new version of ibis-query

**Comment by tswast** [comment]
Still need a fix for query parameters, but making progress.

**Comment by tswast** [comment]
@datapythonista Are we dropping the query parameters feature? I see that `raw_sql` doesn't get `params` or the `QueryContext` that contains the params.

https://github.com/ibis-project/ibis/blob/faa2c368bfd56a34813ceba962bf98540f061312/ibis/backends/base/sql/__init__.py#L91-L131

I think I'll add a `params` argument to `raw_sql` for now and copy over the `execute` from the base class.

**Comment by tswast** [comment]
System tests are now passing locally. I'll do another few checks, but I think we should be good to release ASAP.

**Comment by jreback** [comment]
who hoo!

**Pull request merged**